test_autograd_cuda_gemm (test_na1d.NA1DTests) ... NATTEN failure: cutlass error: Error Internal at: 100

Hi, your work is great! May I ask you a question?
I tried to build NATTEN on V100 but failed. My environment is python=3.8, torch=1.10.1, cuda=11.3.
I bulit NATTEN from source and conducted "make WORKERS=2", but when I conduct "make test", it shows:
 Running unit tests
python -m unittest discover -v -s ./tests
test_autograd_cpu (test_na1d.NA1DTests) ... ok
test_autograd_cuda_gemm (test_na1d.NA1DTests) ... NATTEN failure: cutlass error: Error Internal at: 100

Can you help me find out what's causing this?

Thank you for your reply！Sorry, I couldn't follow up in time due to time difference.

I have run the python snippet as follows：
```
>>> import torch
>>> cuda_device = torch.cuda.get_device_properties(torch.cuda.current_device())
>>> sm = cuda_device.major + cuda_device.minor * 0.1
>>> print(f"{cuda_device=}")
cuda_device=_CudaDeviceProperties(name='Tesla V100-SXM2-16GB', major=7, minor=0, total_memory=16150MB, multi_processor_count=80)
>>> print(f"{sm=}")
sm=7.0
```

When I try to execute the command：`make CUDA_ARCH="7.0"`， it returns：
```
Traceback (most recent call last):
      File "<string>", line 2, in <module>
      File "<pip-setuptools-caller>", line 34, in <module>
      File "/home/featurize/work/NATTEN/setup.py", line 199, in <module>
        setup(
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/__init__.py", line 107, in setup
        return distutils.core.setup(**attrs)
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 185, in setup
        return run_commands(dist)
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
        dist.run_commands()
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 969, in run_commands
        self.run_command(cmd)
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/dist.py", line 1234, in run_command
        super().run_command(command)
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
        cmd_obj.run()
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/command/develop.py", line 34, in run
        self.install_for_development()
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/command/develop.py", line 111, in install_for_development
        self.run_command('build_ext')
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
        self.distribution.run_command(command)
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/dist.py", line 1234, in run_command
        super().run_command(command)
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
        cmd_obj.run()
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/command/build_ext.py", line 84, in run
        _build_ext.run(self)
      File "/home/featurize/work/.local/lib/python3.8/site-packages/Cython/Distutils/old_build_ext.py", line 186, in run
        _build_ext.build_ext.run(self)
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/command/build_ext.py", line 345, in run
        self.build_extensions()
      File "/home/featurize/work/.local/lib/python3.8/site-packages/Cython/Distutils/old_build_ext.py", line 195, in build_extensions
        _build_ext.build_ext.build_extensions(self)
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/command/build_ext.py", line 467, in build_extensions
        self._build_extensions_serial()
      File "/cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/command/build_ext.py", line 493, in _build_extensions_serial
        self.build_extension(ext)
      File "/home/featurize/work/NATTEN/setup.py", line 153, in build_extension
        cuda_arch_list, n_arch = get_cuda_arch_list(cuda_arch) # Expects xx -- i.e. 86
      File "/home/featurize/work/NATTEN/setup.py", line 120, in get_cuda_arch_list
        return _check_cuda_arch(cuda_arch_list), 1
      File "/home/featurize/work/NATTEN/setup.py", line 110, in _check_cuda_arch
        assert arch >= 8.0 and arch < 10, f"Only SM80 and above is supported at this time, got {arch}."
    AssertionError: Only SM80 and above is supported at this time, got 7.0.
    error: subprocess-exited-with-error
    
    × python setup.py develop did not run successfully.
    │ exit code: 1
    ╰─> See above for output.
    
    note: This error originates from a subprocess, and is likely not a problem with pip.
    full command: /cloud/myenv/bin/python -c '
    exec(compile('"'"''"'"''"'"'
    # This is <pip-setuptools-caller> -- a caller that pip uses to run setup.py
    #
    # - It imports setuptools before invoking setup.py, to enable projects that directly
    #   import from `distutils.core` to work with newer packaging standards.
    # - It provides a clear error message when setuptools is not installed.
    # - It sets `sys.argv[0]` to the underlying `setup.py`, when invoking `setup.py` so
    #   setuptools doesn'"'"'t think the script is `-c`. This avoids the following warning:
    #     manifest_maker: standard file '"'"'-c'"'"' not found".
    # - It generates a shim setup.py, for handling setup.cfg-only projects.
    import os, sys, tokenize
    
    try:
        import setuptools
    except ImportError as error:
        print(
            "ERROR: Can not execute `setup.py` since setuptools is not available in "
            "the build environment.",
            file=sys.stderr,
        )
        sys.exit(1)
    
    __file__ = %r
    sys.argv[0] = __file__
    
    if os.path.exists(__file__):
        filename = __file__
        with tokenize.open(__file__) as f:
            setup_py_code = f.read()
    else:
        filename = "<auto-generated setuptools caller>"
        setup_py_code = "from setuptools import setup; setup()"
    
    exec(compile(setup_py_code, filename, "exec"))
    '"'"''"'"''"'"' % ('"'"'/home/featurize/work/NATTEN/setup.py'"'"',), "<pip-setuptools-caller>", "exec"))' develop --no-deps
    cwd: /home/featurize/work/NATTEN/
error: subprocess-exited-with-error
```

So, what should I do to avoid this error?
Looking forward to your reply！

```
7 errors detected in the compilation of "/home/featurize/work/NATTEN/csrc/autogen/src/cuda/naive/source_0.cu".
    make[3]: *** [CMakeFiles/natten.dir/build.make:104: CMakeFiles/natten.dir/autogen/src/cuda/naive/source_0.cu.o] Error 1
    make[3]: *** Waiting for unfinished jobs....
    7 errors detected in the compilation of "/home/featurize/work/NATTEN/csrc/autogen/src/cuda/naive/source_1.cu".
    make[3]: *** [CMakeFiles/natten.dir/build.make:118: CMakeFiles/natten.dir/autogen/src/cuda/naive/source_1.cu.o] Error 1
    make[3]: Leaving directory '/home/featurize/work/NATTEN/build/lib.linux-x86_64-cpython-38'
    make[2]: *** [CMakeFiles/Makefile2:83: CMakeFiles/natten.dir/all] Error 2
    make[2]: Leaving directory '/home/featurize/work/NATTEN/build/lib.linux-x86_64-cpython-38'
    make[1]: *** [Makefile:91: all] Error 2
    make[1]: Leaving directory '/home/featurize/work/NATTEN/build/lib.linux-x86_64-cpython-38'
    /cloud/myenv/lib/python3.8/site-packages/setuptools/dist.py:509: InformationOnly: Normalizing '0.14.7.dev' to '0.14.7.dev0'
      self.metadata.version = self._normalize_version(
    /cloud/myenv/lib/python3.8/site-packages/setuptools/command/develop.py:40: EasyInstallDeprecationWarning: easy_install command is deprecated.
    !!

            ********************************************************************************
            Please avoid running ``setup.py`` and ``easy_install``.
            Instead, use pypa/build, pypa/installer or other
            standards-based tools.

            See https://github.com/pypa/setuptools/issues/917 for details.
            ********************************************************************************

    !!
      easy_install.initialize_options(self)
    /cloud/myenv/lib/python3.8/site-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.
    !!

            ********************************************************************************
            Please avoid running ``setup.py`` directly.
            Instead, use pypa/build, pypa/installer or other
            standards-based tools.

            See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.
            ********************************************************************************

```

--- a/setup.py
+++ b/setup.py
@@ -34,8 +34,12 @@ from setuptools import find_packages, setup
 from setuptools.command.build_ext import build_ext
 from typing import List
 import torch
-from torch.utils.cpp_extension import CUDA_HOME, CppExtension, CUDAExtension, IS_MACOS, IS_WINDOWS, _nt_quote_args, SHARED_FLAG, LIB_EXT
+from torch.utils.cpp_extension import CUDA_HOME, CppExtension, CUDAExtension, _nt_quote_args, SHARED_FLAG, LIB_EXT
 from pathlib import Path
+
+IS_WINDOWS = sys.platform == 'win32'
+IS_MACOS = sys.platform.startswith('darwin')
+IS_LINUX = sys.platform.startswith('linux')
 
 this_directory = Path(__file__).parent
 try:
@@ -179,7 +183,7 @@ class BuildExtension(build_ext):
             # TODO: this is connected to the assertion above; this is wrong, but temporary.
             if current_arch >= 60:
                 cmake_args.append("-DNATTEN_WITH_CUDA_FP16=1")
-            if current_arch >= 60 and CUDA_VERSION >= [11, 0]:
+            if current_arch >= 80 and CUDA_VERSION >= [11, 0]:
                 cmake_args.append("-DNATTEN_WITH_CUDA_BF16=1")
             if current_arch >= 80 and CUDA_VERSION >= [11, 0]:
                 cmake_args.append("-DNATTEN_WITH_CUTLASS=1")

